src/SenseNet.Security/Messaging/SecurityActivityQueue.cs

```diff
@@ -1,597 +1,532 @@
+                SnTrace.SecurityQueue.Write(() => $"SAQT: activity attached to another one: " +
+                                                  $"#SA{activity.Key} -> SA{existing.Key}");
+                existing.Attach(activity);
             }
-
-            internal void Reset()
+            else
+            {
+                SnTrace.SecurityQueue.Write(() =>
+                    $"SAQT: execution ignored immediately: #SA{activity.Key}");
+                activity.StartFinalizationTask();
+            }
+        }
+        private int ExecuteOrChain(SecurityActivity activity, List<SecurityActivity> executingList)
+        {
+            // Discover dependencies
+            foreach (var activityUnderExecution in GetAllFromChains(executingList))
             {
-                // Before call ensure that the arrival queue is empty.
-                lock (_waitingSetLock)
+                if (activity.ShouldWaitFor(activityUnderExecution))
                 {
-                    if (_waitingSet.Count > 0)
-                        SnTrace.SecurityQueue.Write("SAQ: RESET: WaitingSet.Count: {0}", _waitingSet.Count);
-
-                    foreach (var activity in _waitingSet)
-                        activity.Finish();
-                    _waitingSet.Clear();
+                    activity.WaitFor(activityUnderExecution);
+                    //TODO: rewrite SecurityActivityHistoryController and call this: _activityHistory.Wait(activity);
                 }
             }
-            internal void Start()
+
+            // Add to concurrently executable list
+            if (activity.WaitingFor.Count == 0)
             {
-                lock (_waitingSetLock)
-                    _waitingSet.Clear();
+                SnTrace.SecurityQueue.Write(() => $"SAQT: moved to executing list: #SA{activity.Key}");
+                executingList.Add(activity);
             }
-            internal bool IsEmpty => _waitingSet.Count == 0;
 
-            private readonly object _waitingSetLock = new object();
-            private readonly List<SecurityActivity> _waitingSet = new List<SecurityActivity>();
-
-            private bool _run;
-            public void ActivityEnqueued()
+            // Mark as started even if it is waiting.
+            return activity.Id;
+        }
+        private async Task LoadLastActivities(int fromId, CancellationToken cancel)
+        {
+            ConfiguredCancelableAsyncEnumerable<SecurityActivity> loaded;
+            using (var op = SnTrace.SecurityQueue.StartOperation(() => $"DataHandler: LoadLastActivities(fromId: {fromId})"))
             {
-                if (_run)
-                    return;
-                _run = true;
-                Task.Run(ProcessActivities);
+                var loader = new SecurityActivityAsyncLoader(_dataHandler);
+                loaded = loader.LoadAsync(fromId, int.MaxValue, false, cancel).ConfigureAwait(false);
+                op.Successful = true;

```

Comment:
This is not true, the operation has not been finished at this point (we are not waiting for the task here), so the log will contain false info.